### PR TITLE
Separate typechecking from semantics, and improve diagnostics

### LINKED ISSUE #709: Split semantics from TC

### PR DESCRIPTION
**Pull request by kontheocharis**
The typechecking is now independent of the rest of the semantics pass, and lives in a separate crate.

The diagnostics API has been changed slightly to `Diagnostics` and `AccessToDiagnostics` traits. This is useful to access diagnostics methods from a root object, without having to specifically refer to the diagnostics implementor.

Closes #709

### CODE REVIEW

File: compiler/hash-parser/src/diagnostics/warning.rs [hunk 1]
```diff
@@ -60,7 +60,7 @@ impl Display for SubjectKind {
     }
 }
 
-#[derive(Debug)]
+#[derive(Debug, Clone)]
```
Comment (feds01) [inline compiler/hash-parser/src/diagnostics/warning.rs:63]: Why is this needing cloning?

File: compiler/hash-reporting/src/diagnostic.rs [hunk 2]
```diff
@@ -2,50 +2,261 @@
...
+        )*
+    };
+}
+
+make_diagnostic_traits!(Diagnostics with &Self, DiagnosticsMut with &mut Self);
+
+pub struct ImmutableDiagnostics<E, W> {
+    pub errors: RefCell<Vec<E>>,
+    pub warnings: RefCell<Vec<W>>,
+}
```
Comment (feds01) [inline compiler/hash-reporting/src/diagnostic.rs:77]: `@@Docs`, also would be nice to get rid of refcells somehow?

File: compiler/hash-semantics/src/new/ops/context.rs [hunk 3]
```diff
@@ -1,189 +1 @@
...
-        environment::{
-            context::{Binding, BindingKind, BoundVarOrigin, ScopeKind},
-            env::AccessToEnv,
-        },
-        params::{DefParamIndex, ParamId, ParamsId},
-        scopes::StackMemberId,
-    },
-    ty_as_variant,
-};
-use hash_utils::store::{SequenceStore, Store};
```
Comment (feds01) [inline compiler/hash-semantics/src/new/ops/context.rs:1]: File not deleted?

File: compiler/hash-semantics/src/storage/cache.rs [hunk 4]
```diff
@@ -61,12 +61,6 @@ impl<K: Copy + Hash + Eq, V: Clone> PartialStore<K, V> for CacheStore<K, V> {
     fn internal_data(&self) -> &RefCell<HashMap<K, V>> {
         self.store.internal_data()
     }
-
-    /// Clear the [CacheStore] and metrics.
-    fn clear(&self) {
-        self.store.clear();
-        self.reset_metrics();
-    }
```
Comment (feds01) [inline compiler/hash-semantics/src/storage/cache.rs:69]: Was this does not being used before to reset the cache after each module was being walked?

**Review by feds01 — CHANGES_REQUESTED** [review]

File: compiler/hash-reporting/src/diagnostic.rs (hunk 2 above)
Comment (kontheocharis) [reply compiler/hash-reporting/src/diagnostic.rs:77]: Not really possible to get rid of refcells since the access is immutable..

Comment (feds01) [reply compiler/hash-reporting/src/diagnostic.rs:77]: Oh right, the naming is a little confusing, as in it needs to be within an immutable context to append things to the diagnostic store.

Comment (kontheocharis) [reply compiler/hash-reporting/src/diagnostic.rs:77]: I'm open to better naming ideas

Comment (feds01) [reply compiler/hash-reporting/src/diagnostic.rs:89]: Maybe something like `DiagnosticMut` to `Diagnostics`, and `ImmutableDiagnostics` to `CelledDiagnostics`?

File: compiler/hash-semantics/src/storage/cache.rs (hunk 4 above)
Comment (kontheocharis) [reply compiler/hash-semantics/src/storage/cache.rs]: Idk why that was deleted, added it back.

File: compiler/hash-reporting/src/diagnostic.rs (hunk 2 above)
Comment (kontheocharis) [reply compiler/hash-reporting/src/diagnostic.rs:89]: `DiagnosticStore` and `DiagnosticCellStore` ? Cause `Diagnostics` is the name of the trait

**Review by feds01 — APPROVED** [review]

**Pull request merged**